Fixes for macos

hamishmack opened the pull request:
###### Motivation for this change

Gets WebKitGTK+, GtkSourceView and gtk-mac-integration working on MacOS

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [X] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

periklis commented on pkgs/development/libraries/webkitgtk/2.14.nix:
```diff
@@ -18,12 +18,36 @@ stdenv.mkDerivation rec {
+    patch -p0 < ${./PR-153138.patch}
+    patch -p0 < ${./PR-157554.patch}
+    patch -p0 < ${./PR-157574.patch}
+  '';
```
Put these in `patches` and do not use the patch util manually.

periklis commented on pkgs/development/libraries/webkitgtk/2.14.nix:
```diff
@@ -18,12 +18,36 @@ stdenv.mkDerivation rec {
+        Source/WebKit2/CMakeFiles/WebProcess.dir/link.txt
+    sed -i "s|\.\./\.\./lib/libWTFGTK\.a|-Wl,-all_load ../../lib/libWTFGTK.a|" \
+        Source/JavaScriptCore/CMakeFiles/JavaScriptCore.dir/link.txt
+  '';
```
If possible try to use of `substituteInPlace`

periklis commented on pkgs/development/libraries/webkitgtk/2.14.nix:
```diff
@@ -37,24 +61,40 @@ stdenv.mkDerivation rec {
   "-DUSE_LIBHYPHEN=0"
-  "-DENABLE_GLES2=ON"
-  ];
+  ] ++ (if stdenv.isDarwin then [
```
Use of `stdenv.lib.optionals (stdenv.isDarwin) [...]` is a much nicer syntax that avoids parenthesis nesting and plays well with additions for other platforms.

Mic92 replied:
might be an exception here, because of the `else` branch.

periklis commented on pkgs/development/libraries/webkitgtk/2.14.nix:
```diff
@@ -37,24 +61,40 @@ stdenv.mkDerivation rec {
   ] ++ optional enableGeoLocation geoclue2
-    ++ (with xlibs; [ libXdmcp libXt libXtst ]);
+    ++ (with xlibs; [ libXdmcp libXt libXtst ])
+    ++ (if stdenv.isDarwin then [ libedit readline mesa ] else [ wayland ]);
```
As above use `stdenv.lib.optionals` and make the platform explicit for `wayland`.

periklis reviewed (commented):
Thx for the great work, i've proposed some enhnacements.

hamishmack commented:
Ok I think I have the enhancements right now. @Mic92 I used `isLinux` for the `else` case.

LnL7 commented on pkgs/development/libraries/webkitgtk/2.14.nix:
```diff
@@ -32,29 +47,54 @@ stdenv.mkDerivation rec {
-  patches = [ ./finding-harfbuzz-icu.patch ];
+  patches = [ ./finding-harfbuzz-icu.patch ]
+  	++ optionals stdenv.isDarwin [
+    ./PR-152650-2.patch
```
Can you use `fetchpatch` for these?

hamishmack replied:
I don't think we can fetch them directly from macports if that is what you were suggesting.  I had to modify them slightly.  The macports patches are designed to work with `patch -p0`.  I had to add an extra directory to the paths to make them work with `patches` in nix (https://github.com/NixOS/nixpkgs/pull/24700/commits/90f4361fa0dc783c91b8bf2cdf1aefb204ce4893#diff-7e2407d988a384c067352687b99c845e is the commit with that change).

LnL7 replied:
Allrigth

matthewbauer commented:
So if this gets merged I think we can probably close #14284 because it looks like it's roughly the same purpose.

matthewbauer commented:
Also: make sure your commit messages meet [the contributing guideles](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

hamishmack commented:
@matthewbauer I have merged the `Use nix idiomatic patch and replace functions` commit into the first commit and added some detail of on the second commit message.

LnL7 commented:
It's preferred to format the first like `webkgitgtk: fix 2.14 on macOS`. Not mandatory, but it makes it easier for people to read the changelog if everything is formatted consistently.

hamishmack committed:
webkgitgtk: fix 2.14 on macOS

Includes the patches from macports and fixes an issue with the
`otool -L` output that was causing g-ir-scanner to fail.
The paths in the macports patches have been modified so that they
will work with `patches` in nix (an extra directory level was
added).

hamishmack committed:
gtksourceview: fix version 3 on macOS

Add darwin to the list of supported platforms for gtksourceview.

hamishmack committed:
gtk-mac-integration: fix gtk3 version

Include a Gtk3 version of gtk-mac-integration.
Add gobjectIntrospection to gtk-mac-integration.
Use propagatedBuildInputs for gtk.

hamishmack commented:
I have moved the package names to the start of the commit messages.  Split the request into three commits (one for each package).  I also moved the `gtk` from `buildInputs` to `propergatedBuildInputs` in the `gtk-mac-integration` package.
